### ISSUE TITLE
Non-Leader node discards AppendEntries response in raft.h whereas ccfraft.tla does not

### ISSUE THREAD

**Issue by lemmy** [microsoft/CCF#5340]
`raft.h` discards an AppendEntries response if the node is not in the leader state:

https://github.com/microsoft/CCF/blob/671ee7b20e1134633656031666253ed595a0f3cd/src/consensus/aft/raft.h#L1346-L1353

However, no such constraint exist in `ccfraft` causing non-leader nodes to change their `nextIndex` and `matchIndex` when handling an AppendEntries response:

https://github.com/microsoft/CCF/blob/671ee7b20e1134633656031666253ed595a0f3cd/tla/ccfraft.tla#L997-L1008

Ongaro's original TLA+ specification also doesn't constraint handling of AppendEntries responses to leaders.

https://github.com/ongardie/raft.tla/blob/974fff7236545912c035ff8041582864449d0ffe/raft.tla#L392-L402 

Needless to say that Ongaro's original TLA+ `raft.tla` and `ccfraft.tla` violate the following property:

```tla
NextMatchUnchanged ==
    [][\A s \in Server: 
            state[s] # Leader => 
                       /\ nextIndex[s] = nextIndex'[s]
                       /\ matchIndex[s] = matchIndex'[s]]_vars
```

Intuitively, it appears safe to allow nodes to handle AppendEntries responses in any state, as the `nextIndex` and `matchIndex` parameters are not relevant when the node is not the leader, *and*, these variables are reset when the node transitions to the leader state. However, it would be beneficial to provide documentation explaining why the `raft.h` file drops AppendEntries responses in states other than `Leader`, including the `RetiredLeader` state.

A comment in `HandleRequestVoteResponse` makes it explicit that a non-candidate may tally nodes.

https://github.com/microsoft/CCF/blob/671ee7b20e1134633656031666253ed595a0f3cd/tla/ccfraft.tla#L847-L850

**Comment by achamayou** [comment]
> Intuitively, it appears safe to allow nodes to handle AppendEntriesResponses in any state, as the nextIndex and matchIndex parameters are not relevant when the node is not the leader, and, these variables are reset when the node transitions to the leader state. However, it would be beneficial to provide documentation explaining why the raft.h file drops AppendEntries responses in states other than Leader, including the RetiredLeader state.

Neither `nextIndex` nor `matchIndex` are exposed to users, they're strictly internal, and as you point out, they are reset on transition to leader. There is therefore no benefit to computing them on a non-leader, it's just a waste of cycles. I will add a comment to that effect in the code, but I am not (at this time) of the opinion that this should change.

> A comment in HandleRequestVoteResponse makes it explicit that a non-candidate may tally nodes.

This is true in the spec, but not [in the code](https://github.com/microsoft/CCF/blob/f8c31664527f81d6686d3733489a89b4092ee2cb/src/consensus/aft/raft.h#L1625) for the same reason as above.

I suggest in both cases that we align the spec on the behaviour of the code, @heidihoward @jumaffre @eddyashton thoughts?

**Comment by lemmy** [comment]
Aligning the specification with the implementation makes perfect sense to me, especially if we then check the spec for deadlocks and liveness issues.  However, we might want to try and avoid ruling out the possibility of formally showing the refinement of Ongaro's original Raft specification.

**Comment by achamayou** [comment]
@lemmy we are not ruling out making our spec a refinement of Ongaro's, but it's also not a goal, and I don't think we meet it currently. Making the spec accurately model the implementation on the other hand serves our explicit goal of improving the implementation and showing that it is correct.

**Comment by achamayou** [comment]
Resolved by #5356
